Feature/encrypted storage

gsliepen opened the pull request:
This changes the configuration file format to be easier to manipulate in-memory, and optionally allows configuration to either not be stored on disk, or to be encrypted using a key provided by the application.

Commit:
Add support for encrypted storage.

This is a large overhaul of how configuration files are handled. All files
are now in PackMessage format, and are read from disk in to memory in one
go, and also saved to disk from memory in one go, using functions in conf.c.

Commit:
Add the PackMessage library.

Commit:
Add missing declaration of chacha_ivsetup_96().

Commit:
Add support for opening a MeshLink instance without permanent storage.

sairoopsomaraju commented on src/meshlink.c:
```diff
@@ -713,173 +547,91 @@ static bool finalize_join(meshlink_handle_t *mesh) {
+	packmsg_add_str(&out, name);
+	packmsg_add_bin(&out, ecdsa_get_private_key(mesh->private_key), 96);
+	packmsg_add_uint16(&out, atoi(mesh->myport));
 
```
I think writing main config file, writing host config file etc., type of code should be modular. As the loading and unloading of the package message following a format I think we should write it in a function and can reuse in the code in both `meshlink_setup()` and `finalize_join()` which uses the same code.

gsliepen replied:
Good point, I'll add such a function.

sairoopsomaraju commented on src/protocol_edge.c:
```diff
@@ -143,8 +140,10 @@ bool add_edge_h(meshlink_handle_t *mesh, connection_t *c, const char *request) {
 
 	if(!from) {
 		from = new_node();
+		from->status.dirty = true;
```
In many places node `status.dirty` is set to true are we anywhere in the code writing out the configuration of the node and resetting node's dirty status? 
If not so maybe it can be synchronized when meshlink thread has some idle time and if thread is busy with mesh event handling then do it later when it finds idle time.

gsliepen replied:
Hm, that code seems to have gotten lost when I rebased it. It should write it out when the periodic or pingtimeout handler runs. Also, since you now have tests that do `raise(SIGINT)`, I'll check that any updates to node config files that you would expect to have been stored immediately are indeed written out immediately.

sairoopsomaraju commented on src/protocol_auth.c:
```diff
@@ -409,37 +311,32 @@ bool id_h(meshlink_handle_t *mesh, connection_t *c, const char *request) {
 	}
 
-	read_ecdsa_public_key(mesh, c);
+	node_read_public_key(mesh, n);
```
Should the node handle has to be explicitly request for it's public key? While reading configuration don't we store public key it's handle?

gsliepen replied:
No, we only read part of the configuration by default (device class, blacklist status and submesh information). These don't require extra memory to be allocated. Only when we actually want to make a connection (either a meta-connection or UDP) to another node do we read more information from its config file.

Commit:
Various fixes:

- create_initial_config_files() and node_write_config() are now the only
  functions that generate the content of new config files from scratch.
- All public API functions that change config files now immediately
  write them out.
- Config files of nodes that join using an invitation file are immediately
  written out.
- Ensure nodes marked dirty have their config files written out in
  periodic_handler(), and on meshlink_stop().
- Fix some memory leaks.
- Write out updated config files, and recreate mesh->self in meshlink_set_port().

sairoopsomaraju commented on src/conf.c:
```diff
@@ -20,509 +20,549 @@
+		chacha_poly1305_set_key(ctx, mesh->config_key);
+
+		if(len > 12 && chacha_poly1305_decrypt_iv96(ctx, buf, buf + 12, len - 12, decrypted, &decrypted_len)) {
+			free(buf);
```
Memory allocated to `ctx` has to be freed here.

sairoopsomaraju commented on src/meshlink.c:
```diff
@@ -1630,9 +1449,7 @@ void meshlink_close(meshlink_handle_t *mesh) {
 	free(mesh->confbase);
```
Need to free `mesh->private_key` and `mesh->config_key` also while closing

sairoopsomaraju commented on src/meshlink.c:
```diff
@@ -2271,19 +1958,10 @@ bool meshlink_set_port(meshlink_handle_t *mesh, int port) {
-	modify_config_file(mesh, mesh->name, "Port", portstr, true);
 
-	init_configuration(&mesh->config);
+	// TODO: write meshlink.conf again
```
Need to set `mesh->myport` with the new port value and write to the meshlink main config file

sairoopsomaraju commented on src/meshlink.c:
```diff
@@ -2271,19 +1958,10 @@ bool meshlink_set_port(meshlink_handle_t *mesh, int port) {
-	if(!read_server_config(mesh)) {
-		meshlink_errno = MESHLINK_ESTORAGE;
-	} else if(!setup_network(mesh)) {
+	if(!setup_network(mesh)) {
```
I think here we need to create a node for ourself i.e `mesh->self` as it has been freed while `close_network_connections` before `setup_network` or inside `setup_network` before we call `setup_myself` where we use our node handle

gsliepen replied:
Yes, I've added code to recreate `mesh->self` before calling `setup_network()`. Maybe it would be even better if we can prevent `mesh->self` from being freed.

sairoopsomaraju commented on src/meshlink.c:
```diff
@@ -654,18 +512,8 @@ int check_port(meshlink_handle_t *mesh) {
 		int port = 0x1000 + (rand() & 0x7fff);
 
 		if(try_bind(port)) {
```
In `try_bind()` there is a memory leak at `getaddrinfo()` which uses `ai = ai->ai_next` to traverse through addresses and never stored the allocated pointer address.

gsliepen replied:
This was fixed by Gurram's pull request, which I'll merge to master separately.

sairoopsomaraju commented on src/meshlink.c:
```diff
@@ -1105,68 +825,119 @@ static struct timeval idle(event_loop_t *loop, void *data) {
+	if(!node_read_public_key(mesh, mesh->self)) {
+		logger(NULL, MESHLINK_ERROR, "Could not read our host configuration file!");
+		free_node(mesh->self);
+		mesh->self = NULL;
```
All of the host node's config files has to be loaded back into respective node handles before returning. 
I found that we only have `load_node()` and the old function `load_all_nodes()` has been removed somehow.

gsliepen replied:
This is done in `setup_myself()` in `net_setup.c` via the function `config_scan_all(mesh, load_node)`.

sairoopsomaraju commented on src/meshlink.c:
```diff
@@ -1105,68 +825,120 @@ static struct timeval idle(event_loop_t *loop, void *data) {
 static bool meshlink_setup(meshlink_handle_t *mesh) {
-	if(mkdir(mesh->confbase, 0777) && errno != EEXIST) {
-		logger(mesh, MESHLINK_DEBUG, "Could not create directory %s: %s\n", mesh->confbase, strerror(errno));
+	if(!config_init(mesh)) {
```
Unlike `meshlink_read_conf()`, `meshlink_setup()` never called `main_config_lock()` to lock the `meshlink.conf` file.

gsliepen replied:
You're right, it should do that.

sairoopsomaraju commented on src/net_socket.c:
```diff
@@ -704,7 +701,6 @@ void handle_new_meta_connection(event_loop_t *loop, void *data, int flags) {
 
 	c = new_connection();
 	c->name = xstrdup("<unknown>");
-	c->outcompression = mesh->self->connection->outcompression;
```
What does this `outcompression` and `incompression` in node handle and connection handle respectively deal with?

gsliepen replied:
The `incompression` and `outcompression` variables are actually not used anywhere at the moment (I will remove them in a separate commit), but I had to remove this line because `mesh->self->connection` is now always `NULL` in this commit.

sairoopsomaraju commented on doc/Configuration.md:
```diff
@@ -0,0 +1,95 @@
+
+- uint32: invitation format version
+- str: name of the invitee
+- int32: device class of the invitee (may be unused)
```
In `meshlink_invite_ex()` function I also see that there's a `submesh` string parameter while constructing the invitation file. Also I think host file description in this doc should be updated.

sairoopsomaraju commented on src/meshlink.c:
```diff
@@ -2387,66 +2077,40 @@ char *meshlink_invite_ex(meshlink_handle_t *mesh, meshlink_submesh_t *submesh, c
 
-	if(s) {
-		fprintf(f, "SubMesh = %s\n", s->name);
+	for(int i = 0; i < count; i++) {
```
I think adding host config files other than inviter node to the invitation file looks like a TODO item, however I think we should write a condition where if the invitee node belongs to a submesh we should only append host node config files of that submesh only to the invitation.

lakshminarayanagurram commented on src/protocol_auth.c:
```diff
@@ -143,56 +144,35 @@ static bool send_proxyrequest(meshlink_handle_t *mesh, connection_t *c) {
 
-	return send_request(mesh, c, NULL, "%d %s %d.%d %s", ID, mesh->self->connection->name, mesh->self->connection->protocol_major, minor, mesh->appname);
+	return send_request(mesh, c, NULL, "%d %s %d.%d %s", ID, mesh->self->name, PROT_MAJOR, PROT_MINOR, mesh->appname);
 }
```
In meshlink_join, the minor number is hard coded to 1 where as here we are using PROT_MINOR. Shouldn't both requests contain same protocol major and minor number.

sairoopsomaraju commented on src/net_setup.c:
```diff
@@ -612,5 +461,7 @@ void close_network_connections(meshlink_handle_t *mesh) {
 		free(mesh->myport);
```
Shouldn't we be setting `mesh->port = NULL` after freeing it here right?

lakshminarayanagurram commented on src/meshlink.c:
```diff
@@ -1314,6 +1161,17 @@ meshlink_handle_t *meshlink_open_ex(const meshlink_open_params_t *params) {
+
+		if(!prf(params->key, params->keylen, "MeshLink configuration key", 26, mesh->config_key, CHACHA_POLY1305_KEYLEN)) {
+			logger(NULL, MESHLINK_ERROR, "Error creating configuration key!\n");
+			meshlink_errno = MESHLINK_EINTERNAL;
```
In the event of failure, shouldn't we be freeing the `mesh->config_key` and setting it to NULL before returning.

lakshminarayanagurram reviewed (commented):
In the library where ever we allocate and assign handles or string pointers to variables, we don't have any NULL checks for the return values. I think it is useful to add these condition checks where ever we create dynamic memory and assign it to variables in order to detect some critical failures when running meshlink on devices with limited resoures or any other fatal errors.

lakshminarayanagurram commented:
Also i think that the changes from master branch have to be merged into this branch and make changes again to incorporate submesh feature and other fixes present in master and then we have to merge it into master.

sairoopsomaraju commented:
Guus, still writing and reading the host node data into storage is not done properly. When a node joins a mesh and restarts it's instance it's unable to load the other node's address(es) and both the nodes couldn't connect each other. Also MTU probes are only being sent according to logs and of the end is able to respond to it. I'm trying to fix it and will let you know once it is done.

sairoopsomaraju commented on src/conf.c:
```diff
@@ -20,509 +20,551 @@
+	DIR *dir;
+	struct dirent *ent;
+	char dname[PATH_MAX];
+	make_host_path(mesh, NULL, dname, sizeof(dname));
```
We are passing `NULL` as host file name rather than `""` or null character to `make_host_path()` which makes the mesh not to read any of the node files here.

sairoopsomaraju replied:
Guus you missed to this fix in the last push.

sairoopsomaraju commented on src/net.c:
```diff
@@ -589,6 +588,13 @@ static void periodic_handler(event_loop_t *loop, void *data) {
 
+	for splay_each(node_t, n, mesh->nodes) {
+		if(n->status.dirty) {
+			node_write_config(mesh, mesh->self);
```
`node_write_config()` shouldn't be writing `mesh->self` here but the selected dirty node handle

gsliepen replied:
Fixed.

sairoopsomaraju commented on src/meshlink.c:
```diff
@@ -2269,19 +1966,22 @@ bool meshlink_set_port(meshlink_handle_t *mesh, int port) {
-	modify_config_file(mesh, mesh->name, "Port", portstr, true);
+	/* Close down the network. This also deletes mesh->self. */
+	close_network_connections(mesh);
 
```
Storing the new port number into the mesh handle is missing here. Also I think you should consider the order of freeing old port number and allocating port number to `mesh->myport`, writing it out to the storage and closing down the network.

gsliepen replied:
Storing the new port number is done by the call to `write_main_config_files()` right above these lines. But indeed, mesh->myport still needs to be set correctly before calling that.

sairoopsomaraju commented on src/net_setup.c:
```diff
@@ -608,9 +485,10 @@ void close_network_connections(meshlink_handle_t *mesh) {
 	exit_nodes(mesh);
 	exit_connections(mesh);
```
Need to call `exit_submeshes(mesh)`. Also in `exit_submeshes()` function we need to check `mesh->submeshes` parameter and then `list_delete_list(mesh->submeshes)` if the value is not NULL.

gsliepen replied:
Done.

sairoopsomaraju commented on src/net_setup.c:
```diff
@@ -27,326 +27,268 @@
-	fp = fopen(filename, "rb");
+	// While we are at it, read known address information
+	if(!n->canonical_address) {
+		n->canonical_address = packmsg_get_str_dup(&in);
```
In `meshlink_close()` there is no free performed operation of `n->canonical_address` if allocated

gsliepen replied:
Indeed. This should be done in `free_node()`.

sairoopsomaraju commented on src/meshlink.c:
```diff
@@ -674,33 +522,49 @@ int check_port(meshlink_handle_t *mesh) {
+		return false;
+	}
+
+	char *name = packmsg_get_str_dup(&in);
```
Allocated `name` string is not freed after duplicating it to respective node handle(s) parameters

gsliepen replied:
Done.

gsliepen commented:
It now also writes out dirty config files when `meshlink_stop()` is called, and this is also implicitly called by `meshlink_close()`.

Commit:
Add unit test cases for encrypted storage key rotation

Commit:
Modify meshlink configuration base file structre

Commit:
Add encrypted key rotation feature api

Commit:
Ensure we can pass arbitrary binary blobs for keys.

Commit:
Autoconnect to reachable nodes without known public keys

We must allow the autoconnect algorithm to try connections to nodes that
are online but for which we don't have a public key, otherwise we risk
that no connections are formed at all, except to the inviting node.

gsliepen commented:
Merged to master.
